packages/block-library/src/gallery/index.php

```diff
@@ -130,7 +116,58 @@ function block_core_gallery_render( $attributes, $content ) {
 			'context' => 'block-supports',
 		)
 	);
-	return (string) $processed_content;
+
+	// Randomize the order of image blocks. Ideally we should shuffle
+	// the `$parsed_block['innerBlocks']` via the `render_block_data` hook.
+	// However, this hook doesn't apply inner block updates when blocks are
+	// nested.
+	// TODO: In the future, if this hook supports updating innerBlocks in
+	// nested blocks, it should be refactored.
+	// See: https://github.com/WordPress/gutenberg/pull/58733
+	if ( empty( $attributes['randomOrder'] ) ) {
+		return $processed_content;
+	}
+
+	$processor = new WP_HTML_Tag_Processor( (string) $processed_content );
+
+	// Add an index to each figure tag to uniquely identify each Image block.
+	$i = 0;
+	while ( $processor->next_tag( array( 'class_name' => 'wp-block-image' ) ) ) {
+		$processor->set_attribute( 'data-image-index', $i );
+		++$i;
+	}
+	$content = $processor->get_updated_html();
+
+	$pattern = '/<figure[^>]*\bdata-image-index\b[^>]*>.*?<\/figure>/';
+
+	// Find all Image blocks.
+	preg_match_all( $pattern, $content, $matches );
+	if ( ! $matches ) {
+		return $content;
+	}
```

Comment:
```suggestion
	preg_match_all( $pattern, $procesed_content, $matches );
	if ( ! $matches ) {
		return $procesed_content;
	}
```

If we aren't parsing and manipulating the gallery content this can just reference the current markup in `$processed_content`.